File: sdk/typescript/src/organizations.ts
/**
 * Organization management — admin API.
 *
 * Wraps `/api/v1/admin/organizations` plus the user-facing
 * `/api/v1/organizations/invitations/{token}/accept` route.
 *
 * Backend reference: `internal/api/router.go`.
 */

import { httpRequest } from "./http.js";
import { SharkAPIError } from "./errors.js";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/** Organization row as returned by the server. */
export interface Organization {
  id: string;
  name: string;
  slug?: string;
  /** Server stores metadata as a JSON-encoded string. */
  metadata?: string | Record<string, unknown>;
  created_at?: string;
  updated_at?: string;
  [key: string]: unknown;
}

/** Organization member row. */
export interface OrgMember {
  id?: string;
  user_id: string;
  org_id?: string;
  role: string;
  email?: string;
  created_at?: string;
  [key: string]: unknown;
}

/** Organization invitation row. */
export interface OrgInvitation {
  id: string;
  org_id?: string;
  email: string;
  role: string;
  token?: string;
  expires_at?: string;
  created_at?: string;
  [key: string]: unknown;
}

/** Input for {@link OrganizationsClient.create}. */
export interface CreateOrganizationInput {
  name: string;
  slug?: string;
  /** dict will be auto-`JSON.stringify()`-ed; string passed through. */
  metadata?: string | Record<string, unknown>;
  [key: string]: unknown;
}

/** Input for {@link OrganizationsClient.update}. */
export interface UpdateOrganizationInput {
  name?: string;
  slug?: string;
  metadata?: string | Record<string, unknown>;
  [key: string]: unknown;
}

/** Options for {@link OrganizationsClient}. */
export interface OrganizationsClientOptions {
  baseUrl: string;
  adminKey: string;
}

// ---------------------------------------------------------------------------
// Client
// ---------------------------------------------------------------------------

/** Admin client for managing organizations, members, and invitations. */
export class OrganizationsClient {
  private readonly _base: string;
  private readonly _key: string;

  constructor(opts: OrganizationsClientOptions) {
    this._base = opts.baseUrl.replace(/\/+$/, "");
    this._key = opts.adminKey;
  }

  private _auth(): Record<string, string> {
    return { Authorization: `Bearer ${this._key}` };
  }

  private async _throw(status: number, text: string): Promise<never> {
    let code = "api_error";
    let message = text.slice(0, 300);
    try {
      const body = JSON.parse(text) as {
        error?: { code?: string; message?: string } | string;
        message?: string;
      };
      if (typeof body.error === "object" && body.error) {
        if (body.error.code) code = body.error.code;
        if (body.error.message) message = body.error.message;
      } else if (typeof body.error === "string") {
        message = body.error;
      } else if (body.message) {
        message = body.message;
      }
    } catch {
      // keep defaults
    }
    throw new SharkAPIError(message, code, status);
  }

  private static _unwrap<T>(body: unknown): T {
    if (
      body &&
      typeof body === "object" &&
      "data" in (body as Record<string, unknown>) &&
      Object.keys(body as Record<string, unknown>).length <= 2
    ) {
      return (body as { data: T }).data;
    }
    return body as T;
  }

  /** dict input → JSON-encoded string; string passed through. */
  private static _coerceMetadata(value: unknown): string | undefined {
    if (value === undefined || value === null) return undefined;
    if (typeof value === "string") return value;
    return JSON.stringify(value);
  }

  // --------------------------------------------------------------- CRUD

  /** Create an organization (POST /api/v1/admin/organizations). */
  async create(
    name: string,
    slug?: string,
    extra: Record<string, unknown> = {},
  ): Promise<Organization> {
    const body: Record<string, unknown> = { name, ...extra };
    if (slug !== undefined) body.slug = slug;
    if ("metadata" in body) {
      body.metadata = OrganizationsClient._coerceMetadata(body.metadata);
    }
    const url = `${this._base}/api/v1/admin/organizations`;
    const resp = await httpRequest(url, {
      method: "POST",
      headers: { ...this._auth(), "Content-Type": "application/json" },
      body: JSON.stringify(body),
    });
    if (resp.status !== 200 && resp.status !== 201)
      await this._throw(resp.status, resp.text);
    return OrganizationsClient._unwrap<Organization>(resp.json());
  }

  /** List organizations. */
  async list(): Promise<Organization[]> {
    const url = `${this._base}/api/v1/admin/organizations`;
    const resp = await httpRequest(url, { headers: this._auth() });
    if (resp.status !== 200) await this._throw(resp.status, resp.text);
    const body = resp.json<Organization[] | { data?: Organization[] }>();
    if (Array.isArray(body)) return body;
    return body?.data ?? [];
  }

  /** Get a single organization. */
  async get(orgId: string): Promise<Organization> {
    const url = `${this._base}/api/v1/admin/organizations/${encodeURIComponent(orgId)}`;
    const resp = await httpRequest(url, { headers: this._auth() });
    if (resp.status !== 200) await this._throw(resp.status, resp.text);
    return OrganizationsClient._unwrap<Organization>(resp.json());
  }

  /** Update an organization (PATCH). */
  async update(orgId: string, fields: UpdateOrganizationInput): Promise<Organization> {
    const body: Record<string, unknown> = { ...fields };
    if ("metadata" in body) {
      body.metadata = OrganizationsClient._coerceMetadata(body.metadata);
    }
    const url = `${this._base}/api/v1/admin/organizations/${encodeURIComponent(orgId)}`;
    const resp = await httpRequest(url, {
      method: "PATCH",
      headers: { ...this._auth(), "Content-Type": "application/json" },
      body: JSON.stringify(body),
    });
    if (resp.status !== 200) await this._throw(resp.status, resp.text);
    return OrganizationsClient._unwrap<Organization>(resp.json());
  }

  /** Delete an organization. */
  async delete(orgId: string): Promise<void> {
    const url = `${this._base}/api/v1/admin/organizations/${encodeURIComponent(orgId)}`;
    const resp = await httpRequest(url, { method: "DELETE", headers: this._auth() });
    if (resp.status !== 200 && resp.status !== 204)
      await this._throw(resp.status, resp.text);
  }

  // ------------------------------------------------------------- Members

  /** List members of an organization. */
  async listMembers(orgId: string): Promise<OrgMember[]> {
    const url = `${this._base}/api/v1/admin/organizations/${encodeURIComponent(
      orgId,
    )}/members`;
    const resp = await httpRequest(url, { headers: this._auth() });
    if (resp.status !== 200) await this._throw(resp.status, resp.text);
    const body = resp.json<OrgMember[] | { data?: OrgMember[]; members?: OrgMember[] }>();
    if (Array.isArray(body)) return body;
    return body?.data ?? body?.members ?? [];
  }

  /** Add a member directly (best-effort — backend may require invitation flow). */
  async addMember(orgId: string, userId: string, role: string): Promise<OrgMember> {
    const url = `${this._base}/api/v1/admin/organizations/${encodeURIComponent(
      orgId,
    )}/members`;
    const resp = await httpRequest(url, {
      method: "POST",
      headers: { ...this._auth(), "Content-Type": "application/json" },
      body: JSON.stringify({ user_id: userId, role }),
    });
    if (resp.status !== 200 && resp.status !== 201)
      await this._throw(resp.status, resp.text);
    return OrganizationsClient._unwrap<OrgMember>(resp.json());
  }

  /** Update a member's role (PATCH). */
  async updateMemberRole(
    orgId: string,
    memberId: string,
    role: string,
  ): Promise<OrgMember> {
    const url = `${this._base}/api/v1/admin/organizations/${encodeURIComponent(
      orgId,
    )}/members/${encodeURIComponent(memberId)}`;
    const resp = await httpRequest(url, {
      method: "PATCH",
      headers: { ...this._auth(), "Content-Type": "application/json" },
      body: JSON.stringify({ role }),
    });
    if (resp.status !== 200) await this._throw(resp.status, resp.text);
    return OrganizationsClient._unwrap<OrgMember>(resp.json());
  }

  /** Remove a member from the organization. */
  async removeMember(orgId: string, memberId: string): Promise<void> {
    const url = `${this._base}/api/v1/admin/organizations/${encodeURIComponent(
      orgId,
    )}/members/${encodeURIComponent(memberId)}`;
    const resp = await httpRequest(url, { method: "DELETE", headers: this._auth() });
    if (resp.status !== 200 && resp.status !== 204)
      await this._throw(resp.status, resp.text);
  }

  // --------------------------------------------------------- Invitations

  /** List invitations for an organization. */
  async listInvitations(orgId: string): Promise<OrgInvitation[]> {
    const url = `${this._base}/api/v1/admin/organizations/${encodeURIComponent(
      orgId,
    )}/invitations`;
    const resp = await httpRequest(url, { headers: this._auth() });
    if (resp.status !== 200) await this._throw(resp.status, resp.text);
    const body = resp.json<
      OrgInvitation[] | { data?: OrgInvitation[]; invitations?: OrgInvitation[] }
    >();
    if (Array.isArray(body)) return body;
    return body?.data ?? body?.invitations ?? [];
  }

  /** Create a new invitation. */
  async createInvitation(
    orgId: string,
    email: string,
    role: string,
  ): Promise<OrgInvitation> {
    const url = `${this._base}/api/v1/admin/organizations/${encodeURIComponent(
      orgId,
    )}/invitations`;
    const resp = await httpRequest(url, {
      method: "POST",
      headers: { ...this._auth(), "Content-Type": "application/json" },
      body: JSON.stringify({ email, role }),
    });
    if (resp.status !== 200 && resp.status !== 201)
      await this._throw(resp.status, resp.text);
    return OrganizationsClient._unwrap<OrgInvitation>(resp.json());
  }

  /** Delete (cancel) an invitation. */
  async deleteInvitation(orgId: string, invitationId: string): Promise<void> {
    const url = `${this._base}/api/v1/admin/organizations/${encodeURIComponent(
      orgId,
    )}/invitations/${encodeURIComponent(invitationId)}`;
    const resp = await httpRequest(url, { method: "DELETE", headers: this._auth() });
    if (resp.status !== 200 && resp.status !== 204)
      await this._throw(resp.status, resp.text);
  }

  /** Resend an invitation email. */
  async resendInvitation(orgId: string, invitationId: string): Promise<void> {
    const url = `${this._base}/api/v1/admin/organizations/${encodeURIComponent(
      orgId,
    )}/invitations/${encodeURIComponent(invitationId)}/resend`;
    const resp = await httpRequest(url, { method: "POST", headers: this._auth() });
    if (![200, 202, 204].includes(resp.status))
      await this._throw(resp.status, resp.text);
  }

  /**
   * Accept an invitation via token.
   *
   * Calls `POST /api/v1/organizations/invitations/{token}/accept` (the
   * user-facing prefix, NOT `/admin/`). Normally session-cookie auth'd;
   * the admin key header is sent for parity.
   */
  async acceptInvitation(token: string): Promise<Record<string, unknown>> {
    const url = `${this._base}/api/v1/organizations/invitations/${encodeURIComponent(
      token,
    )}/accept`;
    const resp = await httpRequest(url, { method: "POST", headers: this._auth() });
    if (resp.status !== 200 && resp.status !== 201)
      await this._throw(resp.status, resp.text);
    return OrganizationsClient._unwrap<Record<string, unknown>>(resp.json());
  }
}

File: sdk/python/shark_auth/organizations.py
"""Organization management — admin API.

Wraps the admin-key authenticated routes mounted under
``/api/v1/admin/organizations`` plus the user-facing accept-invitation
route at ``/api/v1/organizations/invitations/{token}/accept``.

Backend reference: ``internal/api/router.go``.
"""

from __future__ import annotations

import json
from typing import Any, Dict, List, Optional

from . import _http
from .proxy_rules import _raise


class OrganizationsClient:
    """Admin client for managing organizations, members, and invitations.

    Parameters
    ----------
    base_url:
        Base URL of the SharkAuth server.
    admin_api_key:
        Admin API key (``sk_live_...``).
    session:
        Optional pre-configured :class:`requests.Session`.
    """

    def __init__(
        self,
        base_url: str,
        admin_api_key: str,
        *,
        session: Optional[object] = None,
    ) -> None:
        self._base = base_url.rstrip("/")
        self._token = admin_api_key
        self._session = session or _http.new_session()

    # ------------------------------------------------------------------
    # Helpers
    # ------------------------------------------------------------------

    def _auth(self) -> Dict[str, str]:
        return {"Authorization": f"Bearer {self._token}"}

    @staticmethod
    def _unwrap(body: Any) -> Any:
        if isinstance(body, dict) and "data" in body and len(body) <= 2:
            return body["data"]
        return body

    @staticmethod
    def _coerce_metadata(value: Any) -> Any:
        """Backend stores org metadata as a JSON-encoded *string.

        If the caller passes a dict we serialize it; if it's already a
        string we leave it; ``None`` is passed through.
        """
        if value is None or isinstance(value, str):
            return value
        return json.dumps(value)

    # ------------------------------------------------------------------
    # CRUD
    # ------------------------------------------------------------------

    def create(self, name: str, slug: Optional[str] = None, **extra: Any) -> Dict[str, Any]:
        """Create an organization. POST /api/v1/admin/organizations."""
        body: Dict[str, Any] = {"name": name}
        if slug is not None:
            body["slug"] = slug
        if "metadata" in extra:
            extra["metadata"] = self._coerce_metadata(extra["metadata"])
        body.update(extra)
        url = f"{self._base}/api/v1/admin/organizations"
        resp = _http.request(self._session, "POST", url, headers=self._auth(), json=body)
        if resp.status_code in (200, 201):
            return self._unwrap(resp.json())
        _raise(resp)

    def list(self) -> List[Dict[str, Any]]:
        """List organizations. GET /api/v1/admin/organizations."""
        url = f"{self._base}/api/v1/admin/organizations"
        resp = _http.request(self._session, "GET", url, headers=self._auth())
        if resp.status_code == 200:
            body = resp.json()
            if isinstance(body, dict):
                return body.get("data", []) or []
            return body
        _raise(resp)

    def get(self, org_id: str) -> Dict[str, Any]:
        """Get a single organization. GET /api/v1/admin/organizations/{id}."""
        url = f"{self._base}/api/v1/admin/organizations/{org_id}"
        resp = _http.request(self._session, "GET", url, headers=self._auth())
        if resp.status_code == 200:
            return self._unwrap(resp.json())
        _raise(resp)

    def update(self, org_id: str, **fields: Any) -> Dict[str, Any]:
        """Update an organization. PATCH /api/v1/admin/organizations/{id}."""
        if "metadata" in fields:
            fields["metadata"] = self._coerce_metadata(fields["metadata"])
        url = f"{self._base}/api/v1/admin/organizations/{org_id}"
        resp = _http.request(self._session, "PATCH", url, headers=self._auth(), json=fields)
        if resp.status_code == 200:
            return self._unwrap(resp.json())
        _raise(resp)

    def delete(self, org_id: str) -> None:
        """Delete an organization. DELETE /api/v1/admin/organizations/{id}."""
        url = f"{self._base}/api/v1/admin/organizations/{org_id}"
        resp = _http.request(self._session, "DELETE", url, headers=self._auth())
        if resp.status_code in (200, 204):
            return None
        _raise(resp)

    # ------------------------------------------------------------------
    # Members
    # ------------------------------------------------------------------

    def list_members(self, org_id: str) -> List[Dict[str, Any]]:
        """GET /api/v1/admin/organizations/{id}/members."""
        url = f"{self._base}/api/v1/admin/organizations/{org_id}/members"
        resp = _http.request(self._session, "GET", url, headers=self._auth())
        if resp.status_code == 200:
            body = resp.json()
            if isinstance(body, dict):
                return body.get("data", []) or body.get("members", []) or []
            return body
        _raise(resp)

    def add_member(self, org_id: str, user_id: str, role: str) -> Dict[str, Any]:
        """Add a member directly.

        Note
        ----
        The backend admin surface does not currently expose a direct
        ``POST /members`` route — membership is established via
        ``create_invitation`` + ``accept_invitation``. This method is a
        thin convenience that POSTs to ``/admin/organizations/{id}/members``;
        if the route does not exist the server will return 404/405 which
        is surfaced to the caller via :class:`SharkAPIError`.
        """
        url = f"{self._base}/api/v1/admin/organizations/{org_id}/members"
        body = {"user_id": user_id, "role": role}
        resp = _http.request(self._session, "POST", url, headers=self._auth(), json=body)
        if resp.status_code in (200, 201):
            return self._unwrap(resp.json())
        _raise(resp)

    def update_member_role(self, org_id: str, member_id: str, role: str) -> Dict[str, Any]:
        """PATCH /api/v1/admin/organizations/{id}/members/{member_id}."""
        url = f"{self._base}/api/v1/admin/organizations/{org_id}/members/{member_id}"
        resp = _http.request(self._session, "PATCH", url, headers=self._auth(), json={"role": role})
        if resp.status_code == 200:
            return self._unwrap(resp.json())
        _raise(resp)

    def remove_member(self, org_id: str, member_id: str) -> None:
        """DELETE /api/v1/admin/organizations/{id}/members/{member_id}."""
        url = f"{self._base}/api/v1/admin/organizations/{org_id}/members/{member_id}"
        resp = _http.request(self._session, "DELETE", url, headers=self._auth())
        if resp.status_code in (200, 204):
            return None
        _raise(resp)

    # ------------------------------------------------------------------
    # Invitations
    # ------------------------------------------------------------------

    def list_invitations(self, org_id: str) -> List[Dict[str, Any]]:
        """GET /api/v1/admin/organizations/{id}/invitations."""
        url = f"{self._base}/api/v1/admin/organizations/{org_id}/invitations"
        resp = _http.request(self._session, "GET", url, headers=self._auth())
        if resp.status_code == 200:
            body = resp.json()
            if isinstance(body, dict):
                return body.get("data", []) or body.get("invitations", []) or []
            return body
        _raise(resp)

    def create_invitation(self, org_id: str, email: str, role: str) -> Dict[str, Any]:
        """POST /api/v1/admin/organizations/{id}/invitations."""
        url = f"{self._base}/api/v1/admin/organizations/{org_id}/invitations"
        body = {"email": email, "role": role}
        resp = _http.request(self._session, "POST", url, headers=self._auth(), json=body)
        if resp.status_code in (200, 201):
            return self._unwrap(resp.json())
        _raise(resp)

    def delete_invitation(self, org_id: str, invitation_id: str) -> None:
        """DELETE /api/v1/admin/organizations/{id}/invitations/{invitationId}."""
        url = (
            f"{self._base}/api/v1/admin/organizations/{org_id}"
            f"/invitations/{invitation_id}"
        )
        resp = _http.request(self._session, "DELETE", url, headers=self._auth())
        if resp.status_code in (200, 204):
            return None
        _raise(resp)

    def resend_invitation(self, org_id: str, invitation_id: str) -> None:
        """POST /api/v1/admin/organizations/{id}/invitations/{invitationId}/resend."""
        url = (
            f"{self._base}/api/v1/admin/organizations/{org_id}"
            f"/invitations/{invitation_id}/resend"
        )
        resp = _http.request(self._session, "POST", url, headers=self._auth())
        if resp.status_code in (200, 202, 204):
            try:
                return self._unwrap(resp.json()) if resp.text else None
            except Exception:
                return None
        _raise(resp)

    def accept_invitation(self, token: str) -> Dict[str, Any]:
        """Accept an organization invitation.

        Calls ``POST /api/v1/organizations/invitations/{token}/accept`` —
        note the *user-facing* prefix (no ``/admin/``). This route is
        normally session-cookie authenticated; the admin key header is
        sent for parity with the rest of this client. Callers who need
        true session auth should use a session-bearing client.
        """
        url = f"{self._base}/api/v1/organizations/invitations/{token}/accept"
        resp = _http.request(self._session, "POST", url, headers=self._auth())
        if resp.status_code in (200, 201):
            return self._unwrap(resp.json())
        _raise(resp)
